File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.py
def solution(players, m, k):
    T = len(players)
    expire = [0] * (T + k + 1)
    active = 0
    answer = 0

    for t in range(T):
        active -= expire[t]

        required = players[t] // m

        if active < required:
            add = required - active
            active += add
            answer += add
            expire[t + k] += add  

    return answer

File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.java
class Solution {
    public int solution(int[] players, int m, int k) {
        int answer = 0;
        int T = players.length;
        int[] expire = new int[players.length+k];
        int active = 0;
        int required = 0;
        int add = 0;
        
        for (int t=0; t<T; t++) {
            active -= expire[t];

            required = players[t] / m;

            if (active < required) {
                add = required - active;
                active += add;
                answer += add;
                expire[t + k] += add;
            }
        }
        return answer;
    }
}